### PR TITLE
Fix order of mounts under /run

### PR DESCRIPTION
**Pull request by mrunalp**
Since the hooks are called after user mounts under /run are already setup
we re-order the mounts be inserting a tmpfs at /run and then moving the
uesr mounts over it.

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
@rhatdan PTAL. One caveat with this is that this handles the mounts specified by user but not our internal `/run/secrets` mount. Should we just special handle it?

File: src/systemdhook.c [hunk 1]
```diff
@@ -274,6 +274,75 @@ static bool contains_mount(const char **config_mounts, unsigned len, const char
...
+ */
+static int adjust_run(const char *rootfs,
+		 const char **config_mounts,
+		 unsigned len)
+{
+	int rc = -1;
+	for (unsigned i = 0; i < len; i++) {
+		/* Skip on exact match with /run or /run/ */
+		if (!strcmp("/run", config_mounts[i]) || !strcmp("/run/", config_mounts[i])) {
+			continue;
```
Comment (rhatdan) [inline src/systemdhook.c]: Shouldn't we use RUN_PREVIX for /run/?
Shouldn't we just exit if the user mounted over /run?
If the user does the following:
```
docker run -v /local/run:/run fedora 
```
We should not mount tmpfs on /run.

**Comment by rhatdan** [comment]
I thought /run/secrets would be handled, but yes we should just special handle it.  Check if /run/secrets exists and mv it.

**Comment by rhatdan** [comment]
@runcom @shishir-a412ed PTAL

File: src/systemdhook.c (hunk 1 above)
Comment (mrunalp) [inline src/systemdhook.c]: That should be handled as it is caught in the loop around the call. I'll fix to use RUN_PREFIX.

**Comment by mrunalp** [comment]
@rhatdan Yeah, updating the code for moving /run/secrets as well.

File: src/systemdhook.c (hunk 1 above)
Comment (rhatdan) [inline src/systemdhook.c]: In that case this check is not necessary at all correct?

Comment (mrunalp) [inline src/systemdhook.c]: @rhatdan Yes, you are right. I removed the check.

**Comment by mrunalp** [comment]
@rhatdan Pushed update with suggested changes. I'll remove the debugs after the review before it is ready to merge.

**Comment by mrunalp** [comment]
We also need to remove the `/runtmp` directory. I'll add that code.

**Comment by rhatdan** [comment]
Will this cause a diff to the image, even if nothing changed? IE / was updated?  Also this would fail if the rootfs was mounted readonly, say from an NFS share. Does $ROOTFS/dev/shm exists at the time that we are running? Could we create $ROOTFS/dev/shm/runtmp and them mv off of there?

**Comment by rhatdan** [comment]
Another option would be to mount the tmpfs on /mnt, but I guess we have no guarantee that this exists.

**Comment by mrunalp** [comment]
@rhatdan We could use a directory on the host since they are still accessible. How about using `/tmp`?

**Comment by mrunalp** [comment]
We could create a tmpdir under /tmp and then remove it after the mounts are moved off it.

**Comment by mrunalp** [comment]
Pushed fix to use `/tmp/runtmp.XXXXXX` as temporary directory so this doesn't break readonly rootfs.

**Comment by rhatdan** [comment]
Sounds good.

**Comment by rhatdan** [comment]
Cleanup whitespace.

**Comment by rhatdan** [comment]
This is not working for me when I test it.
```
Jul 27 15:17:11 dhcp-10-19-62-196.boston.devel.redhat.com oci-register-machine[10856]: 2016/07/27 15:17:11 Register machine: prestart 87fead536795a60b1a4ba8d628a9106ef134f97
Jul 27 15:17:11 dhcp-10-19-62-196.boston.devel.redhat.com systemd-machined[2772]: New machine 87fead536795a60b1a4ba8d628a9106e.
Jul 27 15:17:11 dhcp-10-19-62-196.boston.devel.redhat.com oci-systemd-hook[10861]: systemdhook <debug>: Mount Label parsed as: system_u:object_r:svirt_sandbox_file_t:s0:c28,
Jul 27 15:17:11 dhcp-10-19-62-196.boston.devel.redhat.com oci-systemd-hook[10861]: systemdhook <error>: Failed to mount tmpfs at /run: No such file or directory
Jul 27 15:17:11 dhcp-10-19-62-196.boston.devel.redhat.com systemd[1]: Stopped docker container 87fead536795a60b1a4ba8d628a9106ef134f97e939ed033f37d363ed5354e60.
```

**Comment by mrunalp** [comment]
@rhatdan Hmm, that indicates that the tmp directory isn't getting created. I have pushed a fix with the message corrected. Could you retry?

**Comment by rhatdan** [comment]
LGTM. Works now, and I see the /run/secrets along with /run/etc

```
docker run -d -v /etc:/run/etc httpd
```

**Comment by mrunalp** [comment]
@rhatdan Thanks for testing. I removed the debugs and pushed update.

**Comment by mrunalp** [comment]
@rhatdan I think you merged after I pushed so should be good :)

**Comment by rhatdan** [comment]
oops

**Pull request merged**